### PR TITLE
 Integrated drawing shapes

### LINKED ISSUE #16: Integrate drawing shapes
Make drawing shapes with touch input possible.

### PR DESCRIPTION
**Pull request by tedzards509**
Should resolve #16. Also updated dependencies.

Implemented the drawing by adding a new interactableSide called drawShapes. In it, a different GestureDetector is used that calls _onXxxxShape() instead of _onXxxxPiece(). It updates and draws an _arrowAvater every time a new square is reached while panning. When released calls an onCompleteShape callback given into BoardData

Also updated the Shape widget to take a Shape object and draw it either as a circle or as an arrow, depending on its attributes and made it look a bit more like the websites shapes (thickness and round caps. Annoying overlap zones though, found no way to make overlapping custom paints overwrite one another). As the widget now takes any Shape object, I made the Shape class non abstract and deleted the Arrow class.

Added an example implementation in which a long press is used to toggle drawing mode, indicated by a scaling of the board. Considered adding an indicator to the top left corner of the board instead but didn't want to clutter the package too much so I'd recommend making indication of mode the duty of wherever a board is implemented.

### CODE REVIEW

**Comment by tedzards509** [comment]
I intend to add an option to board settings that en/disables drawing shapes instead of the interactible side solution I currently have implemented.

**Comment by tedzards509** [comment]
It should in theory be more of a refactor, but would hypothetically make drawing shapes while being able to play possible.

This would make drawing shapes with left click for integration with input devices with secondary buttons more easily implementable.

Also it feels wrong to hijack the interactableSide enum with a playing unrelated setting.

**Comment by tedzards509** [comment]
Refactoring done. Also re-added Arrow class for backwards compatibility and implemented custom line width as a ratio of the square size. I hope it refreshes this PR correctly.

As for the snapToValidMoves, I did not find a way to make it work efficiently. I'll leave the attribute in ShapeData though.

**Comment by veloce** [comment]
I don't think we should change the shape thickness, they're fine as they are. They don't need to look as in the website and we can add color and opacity variations (see analysis PR in mobile repo).

File: lib/src/widgets/shape.dart [hunk 1]
```diff
@@ -3,50 +3,90 @@ import 'package:flutter/widgets.dart';
...
+class ShapeWidget extends StatelessWidget {
+  const ShapeWidget({
     super.key,
-    required this.color,
     required this.size,
     required this.orientation,
-    required this.fromCoord,
-    required this.toCoord,
+    required this.shape,
+    this.widthRatio = 1 / 6.4, // Default pulled from lichess.org
```
Comment (veloce) [inline lib/src/widgets/shape.dart:12]: Please leave the width like it was

File: lib/src/widgets/shape.dart [hunk 2]
```diff
@@ -3,50 +3,90 @@ import 'package:flutter/widgets.dart';
...
+  final Shape shape;
+  final double widthRatio; // Only applies to arrows
 
   @override
   Widget build(BuildContext context) {
     return SizedBox.square(
       dimension: size,
       child: CustomPaint(
-        painter: _ArrowPainter(color, orientation, fromCoord, toCoord),
+        painter: (shape.dest == null || shape.orig == shape.dest) ?
```
Comment (veloce) [inline lib/src/widgets/shape.dart:25]: I'd rather have a `switch` logic instead.

```dart
painter: switch (shape) {
   case: Arrow(color: final c, orig: final o, dest: final d): _ArrowPainter(...)
   ...
}
```

It is now possible with dart 3 switch expression and sealed class.

File: lib/src/widgets/shape.dart [hunk 3]
```diff
@@ -78,12 +118,17 @@ class _ArrowPainter extends CustomPainter {
     final arrowOffset =
         Offset(arrowHeight * math.cos(angle), arrowHeight * math.sin(angle));
 
+    final paint = Paint()
```
Comment (veloce) [inline lib/src/widgets/shape.dart:121]: I'd prefer to keep the arrow painter as it was, such a change is not part of this PR

**Review by veloce — COMMENTED** [review]
I just commented about the code for now but I could not test because of the conflicts.

File: lib/src/models.dart [hunk 4]
```diff
@@ -355,33 +355,40 @@ Role _toRole(String uciLetter) {
...
 
-/// Base class for shapes that can be drawn on the board.
+/// Class for shapes to be drawn on the board. Draws an arrow from orig to dest or a circle if orig == dest
 @immutable
-abstract class Shape {
+class Shape {
   const Shape({
     required this.color,
     required this.orig,
+    this.dest, // Shape is circle if orig == dest or dest == null
```
[the remark this answers is not in the document]
Comment (tedzards509) [reply lib/src/models.dart:364]: Are you okay with Shape not being abstract or should I split it into Arrow and Circle again?

**Comment by tedzards509** [comment]
I've implemented all the comments. The conflicts are also resolved as of now.
